### PR TITLE
Fix xproto with advanced routing logic.

### PR DESCRIPTION
**Pull request by reshke**
Now xproto will begin`s and various set will be executed correctly, without need in shard connection.

### CODE REVIEW

**Review by EinKrebs — CHANGES_REQUESTED** [review]
Half the tests failed

File: router/relay/relay.go [hunk 1]
```diff
@@ -1021,41 +1018,72 @@ func (rst *RelayStateImpl) ProcessExtendedBuffer(cmngr poolmgr.PoolMgr) error {
...
 
-			_, err = rst.DeployPrepStmt(q.PreparedStatement)
-			if err != nil {
+				rst.saveBind.PreparedStatement = fmt.Sprintf("%d", hash)
+				rst.saveBind.ParameterFormatCodes = q.ParameterFormatCodes
+				rst.Client().SetBindParams(q.Parameters)
+				rst.saveBind.ResultFormatCodes = q.ResultFormatCodes
+				rst.saveBind.Parameters = q.Parameters
+
+				// Do not respond here with bind complete, as relay step should do itself
```
Comment (Denchick) [inline router/relay/relay.go:1057]:
```suggestion
				// Do not respond with BindComplete, as the relay step should take care of itself.

```

File: router/relay/relay.go [hunk 2]
```diff
@@ -1080,12 +1108,34 @@ func (rst *RelayStateImpl) ProcessExtendedBuffer(cmngr poolmgr.PoolMgr) error {
...
 				}
-				defer func() {
-					_ = fin()
-				}()
+
+				for _, msg := range unreplied {
+					// https://www.postgresql.org/docs/current/protocol-flow.html
+					switch qq := msg.(type) {
+					case *pgproto3.RowDescription:
+						// send to client
```
Comment (Denchick) [inline router/relay/relay.go:1120]:
```suggestion
						// send to the client
```

File: router/relay/relay.go [hunk 3]
```diff
@@ -1080,12 +1108,34 @@ func (rst *RelayStateImpl) ProcessExtendedBuffer(cmngr poolmgr.PoolMgr) error {
...
+						if err := rst.Client().Send(qq); err != nil {
+							return err
+						}
+					case *pgproto3.NoData:
+						// send to client
```
Comment (Denchick) [inline router/relay/relay.go:1125]:
```suggestion
						// send to the client
```

**Review by Denchick — CHANGES_REQUESTED** [review]

**Review by x4m — APPROVED** [review]

**Review by EinKrebs — APPROVED** [review]

**Review by Denchick — APPROVED** [review]

**Pull request merged**